zuul-core/src/test/java/com/netflix/zuul/netty/filter/ZuulEndPointRunnerTest.java

```diff
@@ -0,0 +1,141 @@
+import com.netflix.zuul.filters.ZuulFilter;
+import com.netflix.zuul.message.Headers;
+import com.netflix.zuul.message.ZuulMessage;
+import com.netflix.zuul.message.http.HttpQueryParams;
+import com.netflix.zuul.message.http.HttpRequestMessage;
+import com.netflix.zuul.message.http.HttpRequestMessageImpl;
+import com.netflix.zuul.message.http.HttpResponseMessage;
+import com.netflix.zuul.message.http.HttpResponseMessageImpl;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.util.concurrent.ImmediateEventExecutor;
+import rx.Observable;
+
+import static com.netflix.zuul.context.CommonContextKeys.NETTY_SERVER_CHANNEL_HANDLER_CONTEXT;
+import static com.netflix.zuul.context.CommonContextKeys.ZUUL_ENDPOINT;
+import static com.netflix.zuul.netty.filter.ZuulEndPointRunner.DEFAULT_ERROR_ENDPOINT;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@RunWith(JUnit4.class)
+public class ZuulEndPointRunnerTest {
+    private static final String BASIC_ENDPOINT = "basicEndpoint";
+    private ZuulEndPointRunner endpointRunner;
+    private FilterUsageNotifier usageNotifier;
+    private FilterLoader filterLoader;
+    private FilterRunner filterRunner;
+    private Registry registry;
+    private HttpRequestMessageImpl request;
+
+    @Before
+    public void beforeEachTest() {
+        usageNotifier = mock(FilterUsageNotifier.class);
+
+        filterLoader = mock(FilterLoader.class);
+        when(filterLoader.getFilterByNameAndType(DEFAULT_ERROR_ENDPOINT.get(), FilterType.ENDPOINT))
+                .thenReturn(new ErrorEndpoint());
+        when(filterLoader.getFilterByNameAndType(BASIC_ENDPOINT, FilterType.ENDPOINT))
+                .thenReturn(new BasicEndpoint());
+
+        filterRunner = mock(FilterRunner.class);
+        registry = new NoopRegistry();
+        endpointRunner = new ZuulEndPointRunner(usageNotifier, filterLoader, filterRunner, registry);
+
+        SessionContext context = new SessionContext();
+        Headers headers = new Headers();
+        ChannelHandlerContext chc = mock(ChannelHandlerContext.class);
+        when(chc.executor()).thenReturn(ImmediateEventExecutor.INSTANCE);
+        context.put(NETTY_SERVER_CHANNEL_HANDLER_CONTEXT, chc);
+        request = new HttpRequestMessageImpl(context, "http", "GET", "/foo/bar", new HttpQueryParams(), headers, "127.0.0.1", "http", 8080, "server123");
+        request.storeInboundRequest();
+    }
+
+    @Test
+    public void nonErrorEndpoint() {
+        request.getContext().setShouldSendErrorResponse(false);
+        request.getContext().setEndpoint(BASIC_ENDPOINT);
```

Comment:
To test for the real flow, or at least part of it, this should really set the `KEY_ENDPOINT` state on the context.
Setting the endpoint directly defeats the purpose of testing the part where the endpoint is looked up by the runner

Reply:
The test now invokes the set method on the context:

```
request.getContext().set("_endpoint", BASIC_ENDPOINT);
```

I could not use the KEY_ENDPOINT variable.  (KEY_ENDPOINT is a private static variable)

Reply:
`setEndpoint` on context should work though?

Reply:
Ack.  This works for me:
```
        request.getContext().setEndpoint(BASIC_ENDPOINT);
```